server/src/main/java/io/crate/replication/logical/MetadataTracker.java

```diff
@@ -193,137 +190,99 @@ private void run() {
             if (LOGGER.isTraceEnabled()) {
                 LOGGER.trace("Poll metadata for subscription {}", subscriptionName);
             }
-            Consumer<Throwable> onError = e -> {
-                e = SQLExceptions.unwrap(e);
-                if (shouldRetry(e) == false) {
-                    var msg = "Tracking of metadata failed for subscription '" + subscriptionName + "'" +
-                        " with unrecoverable error, stop tracking";
-                    LOGGER.error(msg, e);
-                    replicationService.updateSubscriptionState(
+            processSubscription(subscriptionName).whenComplete(countDown);
+        }
+    }
+
+    private CompletableFuture<?> processSubscription(String subscriptionName) {
+        final ClusterState subscriberState = clusterService.state();
+        var subscription = retrieveSubscription(subscriptionName, subscriberState);
+        if (subscription == null) {
+            if (LOGGER.isTraceEnabled()) {
+                LOGGER.trace("Subscription {} not found inside current local cluster state", subscriptionName);
+            }
+            return CompletableFuture.completedFuture(null);
+        }
+        Client client;
+        try {
+            client = remoteClient.apply(subscriptionName);
+        } catch (Exception e) {
+            return CompletableFuture.failedFuture(e);
+        }
+        return getRemoteClusterState(client)
+            .thenCompose(remoteState -> {
+                PublicationsMetadata publicationsMetadata = remoteState.metadata().custom(PublicationsMetadata.TYPE);
+                if (publicationsMetadata == null) {
+                    LOGGER.trace("No publications found on remote cluster.");
+                    return CompletableFuture.completedFuture(false);
```

Comment:
This mirrors the behavior from before, but I wonder if we should short-circuit with an error to trigger a retry. Otherwise it is still continuing with the other thenCompose clauses

Reply:
Does retry means a retry in the current `run` or just continue to complete the `CountdownFutureCallback` and then retry within the next `run` ?

Reply:
So throwing an error here would then be handled in the `exceptionallyCompose` accordingly. This sounds like a good idea to me.

Reply:
I will approve it now, this could be also a follow up.

Reply:
> Does retry means a retry in the current `run` or just continue to complete the `CountdownFutureCallback` and then retry within the next `run` ?

continue with the next scheduled run.

Reply:
So the behavior is more like `skip for this run` rather than retry.